biom/table.py

```diff
@@ -2935,6 +2921,161 @@ def _intersect_id_order(self, a, b):
+        O2	3.0	4.0	5.0	0.0	0.0	0.0	0.0	0.0	0.0
+        O3	0.0	0.0	0.0	6.0	7.0	8.0	0.0	0.0	0.0
+        O4	0.0	0.0	0.0	9.0	10.0	11.0	0.0	0.0	0.0
+        O5	0.0	0.0	0.0	0.0	0.0	0.0	15.0	16.0	17.0
+
+        """
+        # we grow along the opposite axis
+        invaxis = self._invert_axis(axis)
+        if axis == 'sample':
+            dim_getter = itemgetter(1)
+            stack = hstack
+            invstack = vstack
+        else:
+            dim_getter = itemgetter(0)
+            stack = vstack
+            invstack = hstack
+
+        axis_ids = set()
+        invaxis_ids = set()
+        invaxis_metadata = {}
+
+        all_tables = others[:]
+        all_tables.insert(0, self)
+
+        # verify disjoint, and fetch all ids from all tables
+        for table in all_tables:
+            table_axis_ids = table.ids(axis=axis)
+            table_invaxis_order = table.ids(axis=invaxis)
+            table_invaxis = set(table_invaxis_order)
+
+            # test we have disjoint IDs
+            if not axis_ids.isdisjoint(table_axis_ids):
+                raise DisjointIDError("IDs are not disjoint")
+            axis_ids.update(table_axis_ids)
+
+            if set(table_invaxis) - invaxis_ids:
+                for i in (set(table_invaxis) - invaxis_ids):
+                    invaxis_metadata[i] = table.metadata(i, axis=invaxis)
+
+                # add to our perspective all inv axis IDs
+                invaxis_ids.update(table_invaxis)
+
+        invaxis_order = sorted(invaxis_ids)
+
+        # determine what inv axis IDs do not exist per table, and pad and sort
+        # as necessary
+        padded_tables = []
+        for table in all_tables:
+            missing_ids = list(invaxis_ids - set(table.ids(axis=invaxis)))
+
+            if missing_ids:
+                # determine new shape
+                n_invaxis = len(missing_ids)
+                n_axis = len(table.ids(axis=axis))
+                if axis == 'sample':
+                    shape = (n_invaxis, n_axis)
+                else:
+                    shape = (n_axis, n_invaxis)
+
+                # create the padded matrix
+                zerod = csr_matrix(shape)
+                tmp_mat = invstack([table.matrix_data, zerod])
+
+                # resolve invert axis ids and metadata
+                tmp_inv_ids = list(table.ids(axis=invaxis))
+                tmp_inv_ids.extend(missing_ids)
+                tmp_inv_md = table.metadata(axis=invaxis)
+                if tmp_inv_md is None:
+                    tmp_inv_md = [None] * len(table.ids())
+                else:
+                    tmp_inv_md = list(tmp_inv_md)
+                tmp_inv_md.extend([invaxis_metadata[i] for i in missing_ids])
+
+                # resolve axis ids and metadata
+                tmp_ids = list(table.ids(axis=axis))
+                tmp_md = table.metadata(axis=axis)
+
+                # resolve construction based off axis. This really should be
```

Comment:
the resolution off the axis is something which definitely should be a class method and would be nice to do as it would remove this common pattern from the codebase